### ISSUE TITLE
Can't do arithmetic in filter expressions, countif, sumif, or case conditions

### ISSUE THREAD

**Issue by mazameli** [metabase/metabase#13175]
On our own Stats instance (last build Aug 19), trying to do something like `countif( [Subtotal] + [Tax] > 100 )` in the notebook doesn't work. I've tried this with custom columns, filters, and custom expressions, and it's broken in all of those cases.

I've marked this a P1 because it effectively kills many or even most useful cases for custom columns, filters, and expressions.

## Examples

<img width="506" alt="Screen Shot 2020-08-27 at 9 30 57 AM" src="https://user-images.githubusercontent.com/2223916/91469751-a6af0a00-e848-11ea-8505-562a6b45d95b.png">

<img width="556" alt="image" src="https://user-images.githubusercontent.com/2223916/91469656-80896a00-e848-11ea-91df-fba80cd1473f.png">

<img width="534" alt="Screen Shot 2020-08-27 at 9 28 04 AM" src="https://user-images.githubusercontent.com/2223916/91469835-c80ff600-e848-11ea-9353-04113f8f375c.png">

<img width="531" alt="Screen Shot 2020-08-27 at 9 27 49 AM" src="https://user-images.githubusercontent.com/2223916/91469848-cd6d4080-e848-11ea-93c0-4aae903ac2a5.png">

<img width="508" alt="Screen Shot 2020-08-27 at 9 27 02 AM" src="https://user-images.githubusercontent.com/2223916/91469952-efff5980-e848-11ea-8bed-cae47f10da93.png">

## These examples, without mathematic operators, do work

<img width="1189" alt="Screen Shot 2020-08-27 at 9 30 43 AM" src="https://user-images.githubusercontent.com/2223916/91469809-bfb7bb00-e848-11ea-99f8-95c696ad65a8.png">

<img width="552" alt="Screen Shot 2020-08-27 at 9 27 16 AM" src="https://user-images.githubusercontent.com/2223916/91469876-d6f6a880-e848-11ea-9a22-41fd618e0aa7.png">

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by flamber** [comment]
Related #12268 #12225 #12393 and probably a few issues referenced in those.

**Comment by camsaul** [comment]
So I think the issue here is it's not allowing nested expressions in the condition part of the case expression definition:

https://github.com/metabase/metabase/blob/a380c0af73dc42cedd9f3926d950b623502b39fa/frontend/src/metabase/lib/expressions/parser.js#L226-L243

I tried changing `$.boolean` to `$.expression` but that didn't seem to fix it.

**Comment by ariya** [comment]
This problem stems from the fact that the production for `comparisonExpression` never goes through the productions for any arithmetic expressions (starting from `additionExpression`).

The case `CountIf( [Subtotal] > 10 )` works because there is a production shortcut from the function argument (accepting `any` to `comparisonExpression` through a series of grammar rules) as illustrated in the syntax sub-tree for that particular function argument:

![image](https://user-images.githubusercontent.com/7288/102791547-05e6dc00-435c-11eb-91d4-5416719145be.png)

**However**, in the case of `CountIf( ([Subtotal] + [Tax]) > 0 )`, two productions are competing here, either

* `parenthesisExpression`, which tries to match `([Subtotal] + [Tax])`
* `comparisonExpression`. which tries to match `([Subtotal] + [Tax]) > 0`

Because the former is shorter, the latter will not be reached. Hence, the message "expecting EOF but found >".

Note that the comparison production has another relevant bug: `CountIf( 60 < [Discount] )` does not work as it will yield the error message "Expected boolean but found '60'".

Next step: figure out how to tweak the parser (or the grammar thereof) so that a comparison expression takes also into account a possible arithmetic expression.

**Comment by ariya** [comment]
I got some success experimenting with the grammar revamp.

Before, parts of the grammar (in BNF) look like this. `CASE` is going to use `booleanExpression`, while `atomicExpression` will either match `booleanUnaryExpression` or `comparisonExpression`.

```
booleanExpression ::= atomicExpression (BOOLEAN_OPERATOR_BINARY atomicExpression)*

booleanUnaryExpression := BOOLEAN_OPERATOR_UNARY atomicExpression

comparisonExpression :== dimensionExpression FILTER_OPERATOR expression
```

With the modification, a Boolean expression isn't really part of `atomicExpression` anymore as it is being injected right as an expression with the right operand(s), depending on whether it's a unary (logical NOT) or binary (logical AND or logical OR), with the proper precedence. Basically, it becomes:


```
booleanExpression ::=  logicalOrExpression
  
logicalOrExpression ::= logicalAndExpression (OR logicalAndExpression)*

logicalAndExpression ::= relationalExpression (OR relationalExpression)*

relationalExpression ::=
  booleanUnaryExpression |
  booleanUnaryExpression FILTER_OPERATOR booleanUnaryExpression

booleanUnaryExpression ::= logicalNotExpression |  additionExpression

logicalNotExpression ::= NOT additionExpression

additionExpression ::= multiplicationExpression (ADDITIVE_OPERATOR multiplicationExpression)*
```

which will then permit any kind of regular arithmetic (addition, multiplication, etc) within a filter or comparison.

Next step: ensure that all existing tests do not break, adjust the tree visitors at various places to account for this revised grammar.

**Comment by ariya** [comment]
The revamped expression grammar leads to a working validation and suggestion when doing arithmetic with filtering or aggregation. Some examples are shown in the following screenshots (these were previously not working, and now made working with the revamped grammar):

**Custom Column**

![image](https://user-images.githubusercontent.com/7288/103025601-ff489800-4506-11eb-934d-64603e015568.png)

**Filter**

![image](https://user-images.githubusercontent.com/7288/103025675-230bde00-4507-11eb-9bff-f66f7906001e.png)

**Custom Expressions**

![image](https://user-images.githubusercontent.com/7288/103025484-c01a4700-4506-11eb-8a01-3dc27434e6a9.png)

![image](https://user-images.githubusercontent.com/7288/103025415-9fea8800-4506-11eb-90f7-dc0a3e8f4b7d.png)

![image](https://user-images.githubusercontent.com/7288/103025309-63b72780-4506-11eb-8e78-e297106c58f1.png)

Unfortunately (fortunately, depending on your POV) I also discovered that the current approach to determine whether an identifier is a segment or a metric is also incorrect.

Next step: write up the summary of this segment vs metric issue, attempt to come up an alternative resolving mechanism.

**Comment by ariya** [comment]
The root cause of the ambiguous resolving of a metric, a dimension, or a segment is due to the fact that currently, the parser tries to understand the type of each syntax node during the parsing stage. This is fraught with perils, as demonstrated in this example.

In this following aggregation, the identifier, i.e. `[Clearance]`, refers to a segment:

```
COUNTIF( [Clearance] )
```

while in the following one, the identifier, i.e. `[Price]`, refers to a dimension:

```
COUNTIF( [Price] < 100 ) 
```

However, while parsing isn't completed yet, from the local node perspective, it's hard to determine whether an identifier should be treated (and later resolved) as a segment or a dimension. Consider that the parser consumes the token (from the lexer) _up to_ the identifier (3 tokens each):

```
COUNTIF ( [Clearance]
COUNTIF ( [Price] 
```

In each cases, *only after* the subtree is completed, then the identifier can be categorized properly. For the former, it is because the identifier `[Clearance]` represents a child node of a root node (which must have the Boolean type, essentially performing a type coercion) and for the latter, it is because the identifier `[Price]` is an operand of a comparison (a relational expression, with `<` operator).

Here is a more challenging example:

```
COUNTIF( ([DeliveryDays] < 2) AND [HolidayDeals] )
```

The first identifier, `[DeliveryDays]`, is a dimension while the second one, `[HolidayDeals]`, is a segment.

The only correct way to identify the type of each identifier (or to generalize, the type of each syntax node in the parse tree), is to have a separate post-processing step that visits the tree from the root node and coerce the type of each child nodes based on the operation. For instance, a logical AND (as the example above) can only have two or more Boolean operands.

If we depict the syntax tree for the above example (the filter inside `COUNTIF`), it will look like this:

```
booleanExpression
  logicalOrExpression
    logicalAndExpression
      booleanUnaryExpression
        relationalExpression
          additionExpression
            multiplicationExpression
              atomicExpression
                parenthesisExpression
                  expression
                    relationalExpression "<"
                      additionExpression
                        multiplicationExpression
                          atomicExpression
                            dimensionExpression
                                identifierString [DeliveryDays]
                      additionExpression
                        multiplicationExpression
                          atomicExpression
                            numberLiteral [2]
      booleanUnaryExpression
        relationalExpression
          additionExpression
            multiplicationExpression
              atomicExpression
                dimensionExpression
                    identifierString [HolidayDeals]
```

`[DeliveryDays]` can be resolved as a dimension, because if we trace its ancestors, at some point we reach the relational expression (e.g. comparing with the number 2). Meanwhile, `[HolidayDeals]` can be only resolved as a segment because its ancestor node is the right-hand operand of a logical AND.

My next step: add the above extra post-processing step to determine the identifier correctly.

**Comment by ariya** [comment]
This will be fixed by the new forthcoming feature (#14214) which is slated for the next feature release.

**Comment by rlotun** [comment]
closed by https://github.com/metabase/metabase/pull/15061
